csrc/rsa_cipher.cpp

```diff
@@ -29,56 +53,73 @@ JNIEXPORT jint JNICALL Java_com_amazon_corretto_crypto_provider_RsaCipher_cipher
-        java_buffer outBuf = java_buffer::from_array(env, output, outOff, RSA_size(r));
-        int len = 0;
+        java_buffer outBuf = java_buffer::from_array(env, output, outOff, EVP_PKEY_size(ctx->getKey()));
+
+        size_t len = outBuf.len();
 
         {
             jni_borrow in(env, inBuf, "input buffer");
             jni_borrow out(env, outBuf, "output buffer");
 
+            int ret = 0;
             switch (mode) {
             case 2: // Decrypt
             case 4: // Unwrap
-                len = RSA_private_decrypt(inLength, in.data(), out.data(), r, padding);
+                CHECK_OPENSSL(EVP_PKEY_decrypt_init(keyCtx));
+                setPaddingParams(keyCtx, padding, oaepMdPtr, mgfMdPtr);
+                ret = EVP_PKEY_decrypt(keyCtx, out.data(), &len, in.data(), inLength);
                 break;
             case 1: // Encrypt
             case 3: // Wrap
-                len = RSA_public_encrypt(inLength, in.data(), out.data(), r, padding);
+                CHECK_OPENSSL(EVP_PKEY_encrypt_init(keyCtx));
+                setPaddingParams(keyCtx, padding, oaepMdPtr, mgfMdPtr);
+                ret = EVP_PKEY_encrypt(keyCtx, out.data(), &len, in.data(), inLength);
                 break;
             case -1: // Encrypt with a private key, a.k.a. signing
-                len = RSA_private_encrypt(inLength, in.data(), out.data(), r, padding);
+                CHECK_OPENSSL(EVP_PKEY_sign_init(keyCtx));
+                setPaddingParams(keyCtx, padding, oaepMdPtr, mgfMdPtr);
+                ret = EVP_PKEY_sign(keyCtx, out.data(), &len, in.data(), inLength);
                 break;
             case -2: // Decrypt with a public key, a.k.a verification
-                len = RSA_public_decrypt(inLength, in.data(), out.data(), r, padding);
+                CHECK_OPENSSL(EVP_PKEY_verify_recover_init(keyCtx));
+                setPaddingParams(keyCtx, padding, oaepMdPtr, mgfMdPtr);
+                ret = EVP_PKEY_verify_recover(keyCtx, out.data(), &len, in.data(), inLength);
                 break;
             default:
                 throw_java_ex(EX_RUNTIME_CRYPTO, "Unknown cipher mode");
             }
 
-            if (len < 0) {
-                throw_openssl(EX_BADPADDING, "Unknown error");
+            if (ret <= 0) {
+                long err = drainOpensslErrors();
+                if ((err & RSA_R_DATA_TOO_LARGE_FOR_MODULUS)
+                        || (err & RSA_R_PADDING_CHECK_FAILED)
+                        || (err & RSA_R_OAEP_DECODING_ERROR)) {
+                    throw_java_ex(EX_BADPADDING, formatOpensslError(err, "Bad Padding"));
+                } else {
+                    throw_openssl(formatOpensslError(err, "Unexpected exception").c_str());
+                }
             }
         }
 
-        return len;
+        // mask off high order bytes + sign bits to return non-negative (signed) int
+        return (jint) (len & 0x00007fff);
```

Comment:
> If len > outBuf.len() you need to just abort as we've completely stomped on our memory.

call the [libc abort()](https://man7.org/linux/man-pages/man3/abort.3.html) function that raises SIGABRT, or something else?

> Otherwise, len is in a valid range for jint and requires not masking.

how do you know? i`outlen` [is a `size_t` ptr](https://www.openssl.org/docs/man1.0.2/man3/EVP_PKEY_decrypt.html) and there's no documented cap in the OpenSSL docs, so it could be up to 2^64. admittedly i doubt we'd be processing any buffers close to 2^31 (i.e. max jint val, ~2.1GB), but shouldn't we still handle the case where `len` doesn't actually fit in an int?

Reply:
libc `abort()` is fine.

I know that `len` is valid for `jint` because:
1. Java arrays cannot be longer than Java Integer (`jint`) elements long
2. `outBuf.len()`  is the length of a (subset) of a Java array in `len`, and thus must fit in a `jint`.
3. Thus, if we know that `len > 0` and `len <= outBuf.len()` that `len` must be correctly within the range of `jint`.

(We could skip the logic by knowing that the output must fit within a single RSA modulus and that those are never more than a few kilobytes and thus withing `jint` range, but the above is more formal reasoning.)
5.  `size_t len = outBuf.len();`